File: projects/dejavu.cs/Dejavu.Language/ExprParser.cs
using System.Globalization;
using System.Text.Json.Nodes;

namespace Dejavu.Language;

/// <summary>Pratt-style expression parser over a <see cref="CodeToken"/> stream.</summary>
public sealed class ExprParser
{
    readonly string _file;
    readonly int _base;
    readonly List<CodeToken> _tokens;
    int _pos;

    public ExprParser(string input, string? source = null, string file = "template.dejavu", int @base = 0)
    {
        _file = file;
        _base = @base;
        _tokens = CodeLexer.Lex(input, source ?? input, file, @base);
        foreach (var t in _tokens)
        {
            if (t.Kind == CodeTokenKind.CodeEnd)
            {
                throw new ParseError(
                    "unexpected `%>` inside expression",
                    @base + t.Start,
                    t.End - t.Start,
                    file);
            }
        }
    }

    public static JsonNode Parse(string input, string? source = null, string file = "template.dejavu", int @base = 0)
        => new ExprParser(input, source, file, @base).Parse();

    public JsonNode Parse()
    {
        var expr = ParsePipe();
        if (_pos != _tokens.Count)
        {
            var span = PeekSpan();
            throw new ParseError("trailing input in expression", span.start, span.length, _file, "unexpected");
        }
        return expr;
    }

    CodeToken? Peek() => _pos < _tokens.Count ? _tokens[_pos] : null;

    CodeTokenKind? PeekKind() => Peek()?.Kind;

    (int start, int length) PeekSpan()
    {
        if (_pos < _tokens.Count)
        {
            var t = _tokens[_pos];
            return (_base + t.Start, Math.Max(1, t.End - t.Start));
        }
        var end = _base + (_tokens.Count > 0 ? _tokens[^1].End : 0);
        return (end, 1);
    }

    CodeToken? Bump() => _pos < _tokens.Count ? _tokens[_pos++] : null;

    string ExpectIdent()
    {
        var t = Bump();
        if (t is { Kind: CodeTokenKind.Ident } tok) return tok.Text;
        var span = PeekSpan();
        throw new ParseError("expected identifier", span.start, span.length, _file, "expected ident");
    }

    JsonNode ParsePipe()
    {
        var left = ParseOr();
        while (PeekKind() == CodeTokenKind.PipeOp)
        {
            Bump();
            var filter = ExpectIdent();
            var args = new JsonArray();
            if (PeekKind() == CodeTokenKind.LParen)
            {
                Bump();
                if (PeekKind() != CodeTokenKind.RParen)
                {
                    while (true)
                    {
                        args.Add(ParsePipe());
                        if (PeekKind() == CodeTokenKind.Comma)
                        {
                            Bump();
                            continue;
                        }
                        break;
                    }
                }
                if (Bump()?.Kind != CodeTokenKind.RParen)
                {
                    var span = PeekSpan();
                    throw new ParseError("expected `)` after filter arguments", span.start, span.length, _file);
                }
            }
            left = new JsonObject
            {
                ["type"] = "Expr.Pipe",
                ["expression"] = left,
                ["filter"] = filter,
                ["arguments"] = args,
            };
        }
        return left;
    }

    JsonNode ParseOr()
    {
        var left = ParseAnd();
        while (PeekKind() == CodeTokenKind.OrOr)
        {
            Bump();
            left = Bin("||", left, ParseAnd());
        }
        return left;
    }

    JsonNode ParseAnd()
    {
        var left = ParseCmp();
        while (PeekKind() == CodeTokenKind.AndAnd)
        {
            Bump();
            left = Bin("&&", left, ParseCmp());
        }
        return left;
    }

    JsonNode ParseCmp()
    {
        var left = ParseAdd();
        var op = PeekKind() switch
        {
            CodeTokenKind.EqEq => "==",
            CodeTokenKind.NotEq => "!=",
            CodeTokenKind.LessEq => "<=",
            CodeTokenKind.GreaterEq => ">=",
            CodeTokenKind.Less => "<",
            CodeTokenKind.Greater => ">",
            CodeTokenKind.In => "in",
            _ => null,
        };
        if (op is not null)
        {
            Bump();
            return Bin(op, left, ParseAdd());
        }
        return left;
    }

    JsonNode ParseAdd()
    {
        var left = ParseMul();
        while (true)
        {
            var kind = PeekKind();
            if (kind is CodeTokenKind.Plus or CodeTokenKind.Minus)
            {
                var op = kind == CodeTokenKind.Plus ? "+" : "-";
                Bump();
                left = Bin(op, left, ParseMul());
            }
            else break;
        }
        return left;
    }

    JsonNode ParseMul()
    {
        var left = ParseUnary();
        while (true)
        {
            var kind = PeekKind();
            if (kind is CodeTokenKind.Star or CodeTokenKind.Slash or CodeTokenKind.Percent)
            {
                var op = kind == CodeTokenKind.Star ? "*" : kind == CodeTokenKind.Slash ? "/" : "%";
                Bump();
                left = Bin(op, left, ParseUnary());
            }
            else break;
        }
        return left;
    }

    JsonNode ParseUnary()
    {
        var kind = PeekKind();
        if (kind is CodeTokenKind.Bang or CodeTokenKind.Minus or CodeTokenKind.Plus)
        {
            var op = kind == CodeTokenKind.Bang ? "!" : kind == CodeTokenKind.Minus ? "-" : "+";
            Bump();
            return new JsonObject
            {
                ["type"] = "Expr.Unary",
                ["operator"] = op,
                ["argument"] = ParseUnary(),
            };
        }
        return ParsePostfix();
    }

    JsonNode ParsePostfix()
    {
        var left = ParsePrimary();
        while (true)
        {
            var kind = PeekKind();
            if (kind == CodeTokenKind.Dot)
            {
                Bump();
                left = new JsonObject
                {
                    ["type"] = "Expr.Member",
                    ["object"] = left,
                    ["property"] = ExpectIdent(),
                };
            }
            else if (kind == CodeTokenKind.LBracket)
            {
                Bump();
                var index = ParsePipe();
                if (Bump()?.Kind != CodeTokenKind.RBracket)
                {
                    var span = PeekSpan();
                    throw new ParseError("expected `]`", span.start, span.length, _file);
                }
                left = new JsonObject
                {
                    ["type"] = "Expr.Index",
                    ["object"] = left,
                    ["index"] = index,
                };
            }
            else if (kind == CodeTokenKind.LParen)
            {
                Bump();
                var args = new JsonArray();
                if (PeekKind() != CodeTokenKind.RParen)
                {
                    while (true)
                    {
                        args.Add(ParsePipe());
                        if (PeekKind() == CodeTokenKind.Comma)
                        {
                            Bump();
                            continue;
                        }
                        break;
                    }
                }
                if (Bump()?.Kind != CodeTokenKind.RParen)
                {
                    var span = PeekSpan();
                    throw new ParseError("expected `)`", span.start, span.length, _file);
                }
                left = new JsonObject
                {
                    ["type"] = "Expr.Call",
                    ["callee"] = left,
                    ["arguments"] = args,
                };
            }
            else break;
        }
        return left;
    }

    JsonNode ParsePrimary()
    {
        var t = Bump();
        if (t is null)
        {
            var span = PeekSpan();
            throw new ParseError("unexpected end of expression", span.start, span.length, _file);
        }

        switch (t.Value.Kind)
        {
            case CodeTokenKind.String:
                return new JsonObject { ["type"] = "Expr.Literal", ["value"] = t.Value.Text };
            case CodeTokenKind.Bool:
                return new JsonObject { ["type"] = "Expr.Literal", ["value"] = t.Value.Value == true };
            case CodeTokenKind.Null:
                return new JsonObject { ["type"] = "Expr.Literal", ["value"] = null };
            case CodeTokenKind.Number:
            {
                if (!double.TryParse(t.Value.Text, NumberStyles.Float, CultureInfo.InvariantCulture, out var num))
                {
                    throw new ParseError(
                        $"invalid number `{t.Value.Text}`",
                        _base + t.Value.Start,
                        t.Value.End - t.Value.Start,
                        _file);
                }
                return new JsonObject { ["type"] = "Expr.Literal", ["value"] = num };
            }
            case CodeTokenKind.Ident:
                return new JsonObject { ["type"] = "Expr.Identifier", ["name"] = t.Value.Text };
            case CodeTokenKind.LParen:
            {
                var e = ParsePipe();
                if (Bump()?.Kind != CodeTokenKind.RParen)
                {
                    var span = PeekSpan();
                    throw new ParseError("expected `)`", span.start, span.length, _file);
                }
                return e;
            }
            default:
                throw new ParseError(
                    "unexpected token in expression",
                    _base + t.Value.Start,
                    Math.Max(1, t.Value.End - t.Value.Start),
                    _file);
        }
    }

    static JsonObject Bin(string op, JsonNode left, JsonNode right) => new()
    {
        ["type"] = "Expr.Binary",
        ["operator"] = op,
        ["left"] = left,
        ["right"] = right,
    };
}

File: projects/dejavu.kt/dejavu-language/src/main/kotlin/dejavu/language/ExprParser.kt
package dejavu.language

import kotlinx.serialization.json.*

/** Pratt-style expression parser over a [CodeToken] stream. */
class ExprParser(
    private val source: String,
    private val file: String,
    private val base: Int,
    input: String,
) {
    private val tokens: List<CodeToken> = CodeLexer.lexCode(input, source = source, file = file, base = base)
    private var pos = 0

    init {
        for (t in tokens) {
            if (t.kind == CodeTokenKind.CodeEnd) {
                throw ParseError(
                    message = "unexpected `%>` inside expression",
                    file = file,
                    start = base + t.start,
                    length = t.end - t.start,
                )
            }
        }
    }

    fun parse(): JsonObject {
        val expr = parsePipe()
        if (pos != tokens.size) {
            val span = peekSpan()
            throw ParseError(
                message = "trailing input in expression",
                file = file,
                start = span.first,
                length = span.second,
                label = "unexpected",
            )
        }
        return expr
    }

    private fun peek(): CodeToken? = tokens.getOrNull(pos)

    private fun peekKind(): CodeTokenKind? = tokens.getOrNull(pos)?.kind

    private fun peekSpan(): Pair<Int, Int> {
        val t = tokens.getOrNull(pos)
        if (t != null) return (base + t.start) to maxOf(1, t.end - t.start)
        val last = tokens.lastOrNull()
        val end = base + (last?.end ?: 0)
        return end to 1
    }

    private fun bump(): CodeToken? = tokens.getOrNull(pos++)

    private fun expectIdent(): String {
        val t = bump()
        if (t?.kind == CodeTokenKind.Ident) return t.text
        val span = peekSpan()
        throw ParseError(
            message = "expected identifier",
            file = file,
            start = span.first,
            length = span.second,
            label = "expected ident",
        )
    }

    private fun parsePipe(): JsonObject {
        var left = parseOr()
        while (peekKind() == CodeTokenKind.PipeOp) {
            bump()
            val filter = expectIdent()
            val args = mutableListOf<JsonElement>()
            if (peekKind() == CodeTokenKind.LParen) {
                bump()
                if (peekKind() != CodeTokenKind.RParen) {
                    while (true) {
                        args += parsePipe()
                        if (peekKind() == CodeTokenKind.Comma) {
                            bump()
                            continue
                        }
                        break
                    }
                }
                if (bump()?.kind != CodeTokenKind.RParen) {
                    val span = peekSpan()
                    throw ParseError(
                        message = "expected `)` after filter arguments",
                        file = file,
                        start = span.first,
                        length = span.second,
                    )
                }
            }
            left = buildJsonObject {
                put("type", "Expr.Pipe")
                put("expression", left)
                put("filter", filter)
                put("arguments", JsonArray(args))
            }
        }
        return left
    }

    private fun parseOr(): JsonObject {
        var left = parseAnd()
        while (peekKind() == CodeTokenKind.OrOr) {
            bump()
            left = bin("||", left, parseAnd())
        }
        return left
    }

    private fun parseAnd(): JsonObject {
        var left = parseCmp()
        while (peekKind() == CodeTokenKind.AndAnd) {
            bump()
            left = bin("&&", left, parseCmp())
        }
        return left
    }

    private fun parseCmp(): JsonObject {
        val left = parseAdd()
        val op = when (peekKind()) {
            CodeTokenKind.EqEq -> "=="
            CodeTokenKind.NotEq -> "!="
            CodeTokenKind.LessEq -> "<="
            CodeTokenKind.GreaterEq -> ">="
            CodeTokenKind.Less -> "<"
            CodeTokenKind.Greater -> ">"
            CodeTokenKind.In -> "in"
            else -> null
        }
        if (op != null) {
            bump()
            return bin(op, left, parseAdd())
        }
        return left
    }

    private fun parseAdd(): JsonObject {
        var left = parseMul()
        while (true) {
            val kind = peekKind()
            if (kind == CodeTokenKind.Plus || kind == CodeTokenKind.Minus) {
                val op = if (kind == CodeTokenKind.Plus) "+" else "-"
                bump()
                left = bin(op, left, parseMul())
            } else break
        }
        return left
    }

    private fun parseMul(): JsonObject {
        var left = parseUnary()
        while (true) {
            val kind = peekKind()
            if (kind == CodeTokenKind.Star || kind == CodeTokenKind.Slash || kind == CodeTokenKind.Percent) {
                val op = when (kind) {
                    CodeTokenKind.Star -> "*"
                    CodeTokenKind.Slash -> "/"
                    else -> "%"
                }
                bump()
                left = bin(op, left, parseUnary())
            } else break
        }
        return left
    }

    private fun parseUnary(): JsonObject {
        val kind = peekKind()
        if (kind == CodeTokenKind.Bang || kind == CodeTokenKind.Minus || kind == CodeTokenKind.Plus) {
            val op = when (kind) {
                CodeTokenKind.Bang -> "!"
                CodeTokenKind.Minus -> "-"
                else -> "+"
            }
            bump()
            return buildJsonObject {
                put("type", "Expr.Unary")
                put("operator", op)
                put("argument", parseUnary())
            }
        }
        return parsePostfix()
    }

    private fun parsePostfix(): JsonObject {
        var left = parsePrimary()
        while (true) {
            when (peekKind()) {
                CodeTokenKind.Dot -> {
                    bump()
                    left = buildJsonObject {
                        put("type", "Expr.Member")
                        put("object", left)
                        put("property", expectIdent())
                    }
                }

                CodeTokenKind.LBracket -> {
                    bump()
                    val index = parsePipe()
                    if (bump()?.kind != CodeTokenKind.RBracket) {
                        val span = peekSpan()
                        throw ParseError(
                            message = "expected `]`",
                            file = file,
                            start = span.first,
                            length = span.second,
                        )
                    }
                    left = buildJsonObject {
                        put("type", "Expr.Index")
                        put("object", left)
                        put("index", index)
                    }
                }

                CodeTokenKind.LParen -> {
                    bump()
                    val args = mutableListOf<JsonElement>()
                    if (peekKind() != CodeTokenKind.RParen) {
                        while (true) {
                            args += parsePipe()
                            if (peekKind() == CodeTokenKind.Comma) {
                                bump()
                                continue
                            }
                            break
                        }
                    }
                    if (bump()?.kind != CodeTokenKind.RParen) {
                        val span = peekSpan()
                        throw ParseError(
                            message = "expected `)`",
                            file = file,
                            start = span.first,
                            length = span.second,
                        )
                    }
                    left = buildJsonObject {
                        put("type", "Expr.Call")
                        put("callee", left)
                        put("arguments", JsonArray(args))
                    }
                }

                else -> break
            }
        }
        return left
    }

    private fun parsePrimary(): JsonObject {
        val t = bump()
        if (t == null) {
            val span = peekSpan()
            throw ParseError(
                message = "unexpected end of expression",
                file = file,
                start = span.first,
                length = span.second,
            )
        }
        return when (t.kind) {
            CodeTokenKind.String -> buildJsonObject {
                put("type", "Expr.Literal")
                put("value", t.text)
            }

            CodeTokenKind.Bool -> buildJsonObject {
                put("type", "Expr.Literal")
                put("value", t.value == true)
            }

            CodeTokenKind.Null -> buildJsonObject {
                put("type", "Expr.Literal")
                put("value", JsonNull)
            }

            CodeTokenKind.Number -> {
                val num = t.text.toDoubleOrNull()
                    ?: throw ParseError(
                        message = "invalid number `${t.text}`",
                        file = file,
                        start = base + t.start,
                        length = t.end - t.start,
                    )
                buildJsonObject {
                    put("type", "Expr.Literal")
                    if (num % 1.0 == 0.0) put("value", num.toLong())
                    else put("value", num)
                }
            }

            CodeTokenKind.Ident -> buildJsonObject {
                put("type", "Expr.Identifier")
                put("name", t.text)
            }

            CodeTokenKind.LParen -> {
                val e = parsePipe()
                if (bump()?.kind != CodeTokenKind.RParen) {
                    val span = peekSpan()
                    throw ParseError(
                        message = "expected `)`",
                        file = file,
                        start = span.first,
                        length = span.second,
                    )
                }
                e
            }

            else -> throw ParseError(
                message = "unexpected token in expression",
                file = file,
                start = base + t.start,
                length = maxOf(1, t.end - t.start),
            )
        }
    }

    private fun bin(op: String, left: JsonObject, right: JsonObject) = buildJsonObject {
        put("type", "Expr.Binary")
        put("operator", op)
        put("left", left)
        put("right", right)
    }

    companion object {
        fun parse(
            input: String,
            source: String = input,
            file: String = "template.dejavu",
            base: Int = 0,
        ): JsonObject = ExprParser(source, file, base, input).parse()
    }
}

fun parseExpr(
    input: String,
    source: String = input,
    file: String = "template.dejavu",
    base: Int = 0,
): JsonObject = ExprParser.parse(input, source = source, file = file, base = base)
